File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int dist[20001];

vector<int> adj[20001];

int solution(int n, vector<vector<int>> edge) {
    int answer = 0;

    // 간선 연결 리스트
    for(int i = 0; i < edge.size(); ++i){
        adj[edge[i][0]].push_back(edge[i][1]);
        adj[edge[i][1]].push_back(edge[i][0]);
    }
    
    // 초기화
    for(int i = 0; i < 20001; ++i) dist[i] = -1;
    
    queue<int> q;
    q.push(1);
    dist[1] = 0;
    
    int max_dist = 0;
    
    while(!q.empty()){
        int cur = q.front();
        q.pop();
        
        if(dist[cur] > max_dist) max_dist = dist[cur];
        
        for(int i = 0; i < adj[cur].size(); ++i) {
            int next = adj[cur][i];
            if(dist[next] == -1) {
                dist[next] = dist[cur] + 1;
                q.push(next);
            }
        }
    }
    
    for(int i = 1; i <= n; ++i){
        if(dist[i] == max_dist) answer++;
    }
    
    return answer;
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    
    int[] dist;
    List<Integer>[] adjList;
    
    public int solution(int n, int[][] edge) {
        int answer = 0;
        
        // init
        dist = new int[n + 1];
        Arrays.fill(dist, -1);
        adjList = new ArrayList[n + 1];
        for(int i = 0; i < n + 1; i++) {
            adjList[i] = new ArrayList<>();
        }
        
        // edge -> adgList
        for(int i = 0; i < edge.length; i++) {
            adjList[edge[i][0]].add(edge[i][1]);
            adjList[edge[i][1]].add(edge[i][0]);
        }        
        
        // bfs
        Queue<Integer> q = new ArrayDeque<>();
        q.offer(1);
        dist[1] = 0;
        
        while(!q.isEmpty()) {
            int cur = q.poll();
            
            for(int i = 0; i < adjList[cur].size(); i++) {
                int next = adjList[cur].get(i);
                
                if(dist[next] != -1) continue;
                
                dist[next] = dist[cur] + 1;
                q.offer(next);
            }
        }
        
        // max dist
        int max_dist = 0;
        for(int i = 1; i <= n; i++) {
            if(dist[i] > max_dist) max_dist = dist[i];
        }
        
        // count
        for(int i = 1; i <= n; i++) {
            if(dist[i] == max_dist) answer++;
        }
        
        return answer;
    }
}